### PR TITLE
repair create database error:Couldn't find table xxxx in database xxxx

### PR DESCRIPTION
**Pull request by leehand**
https://github.com/zendesk/maxwell/issues/1899

### CODE REVIEW

**Comment by osheroff** [comment]
Hi, no, sorry; this is completely wrong.  The solution to your problem is to add the table you can't read into the blacklist.  So:

```
bin/maxwell --filter="blacklist: bad_db.bad_table"
bin/maxwell --filter="blacklist: bad_db.*"
```

**Comment by leehand** [comment]
I know, but in reality a mysql will be used by many departments or users. You often don't know the database they created. You can only know that they have created the database  after the error.

**Comment by leehand** [comment]
Moreover, in terms of whitelist and blacklist, it is more appropriate to take the intersection.

**Comment by osheroff** [comment]
unfortunately if I merged this, maxwell would never learn about tables not in a whitelist, so a whitelist would be locked in -- in order words, Maxwell would never learn about new databases/tables and you couldn't add them to the whitelist.

**Comment by osheroff** [comment]
I would accept a patch that added a `--ignore-missing-schema` flag for configurations like yours; if this flag was set we'd rescue and continue with missing tables

**Comment by leehand** [comment]
Thanks for your patient reply, did you mean to add --ignore-missing-schema：true/false  ?
Ignore database and table synchronization not configured in filter-configuration:
image

**Comment by osheroff** [comment]
no, i meant more like:
"if `--ignore-missing-schema` is set, try to continue on when we encounter missing tables or databases".   I don't know what that looks like code-wise right now, and am leaving for the week to go sit in the woods in a tent.  cheers!

**Comment by leehand** [comment]
Okay, when you come back, please take a look at the information. I have added an ignore_missing_schema configuration to the config.properties file：

#Ignore database and table schema add not configured in filter-configuration. Default false
#ignore_missing_schema = false


And the following judgment will be made before throwing an exception ：
  if (!ignoreMissingSchema ||  (ignoreMissingSchema && filter.includes(dbName, tblName)))
image

**Comment by xcvecs** [comment]
> Okay, when you come back, please take a look at the information. I have added an ignore_missing_schema configuration to the config.properties file：
> 
> #Ignore database and table schema add not configured in filter-configuration. Default false #ignore_missing_schema = false
> 
> And the following judgment will be made before throwing an exception ： if (!ignoreMissingSchema || (ignoreMissingSchema && filter.includes(dbName, tblName))) image

why add picture?

**Comment by leehand** [comment]
> > Okay, when you come back, please take a look at the information. I have added an ignore_missing_schema configuration to the config.properties file：
> > #Ignore database and table schema add not configured in filter-configuration. Default false #ignore_missing_schema = false
> > And the following judgment will be made before throwing an exception ： if (!ignoreMissingSchema || (ignoreMissingSchema && filter.includes(dbName, tblName))) image
> 
> why add picture?

sorry, I don't quite understand what you mean.

File: src/main/java/com/zendesk/maxwell/replication/TableCache.java [hunk 1]
```diff
@@ -15,22 +15,28 @@ public TableCache(String maxwellDB) {
...
 				return;
 			}
 
+
 			Database db = schema.findDatabase(dbName);
-			if ( db == null )
-				throw new RuntimeException("Couldn't find database " + dbName);
-			else {
+			if ( db == null ) {
+				if (!ignoreMissingSchema ||  (ignoreMissingSchema && filter.includes(dbName, tblName)))
```
Comment (xcvecs) [inline src/main/java/com/zendesk/maxwell/replication/TableCache.java:27]: add comment

**Review by xcvecs — COMMENTED** [review]
add comment

File: src/main/java/com/zendesk/maxwell/replication/TableCache.java (hunk 1 above)
Comment (xcvecs) [inline src/main/java/com/zendesk/maxwell/replication/TableCache.java:27]: add fix

File: src/main/java/com/zendesk/maxwell/replication/TableCache.java [hunk 2]
```diff
@@ -15,22 +15,28 @@ public TableCache(String maxwellDB) {
...
+					throw new RuntimeException("Couldn't find database " + dbName);
```
Comment (xcvecs) [inline src/main/java/com/zendesk/maxwell/replication/TableCache.java:28]: add fix

File: src/main/java/com/zendesk/maxwell/replication/TableCache.java [hunk 3]
```diff
@@ -15,22 +15,28 @@ public TableCache(String maxwellDB) {
...
+
+			}else {
 				Table tbl = db.findTable(tblName);
 
-				if (tbl == null)
-					throw new RuntimeException("Couldn't find table " + tblName + " in database " + dbName);
-				else
+				if (tbl == null) {
+					if (!ignoreMissingSchema ||  (ignoreMissingSchema && filter.includes(dbName, tblName)))
```
Comment (xcvecs) [inline src/main/java/com/zendesk/maxwell/replication/TableCache.java:34]: add comment or review.

File: src/main/java/com/zendesk/maxwell/replication/TableCache.java [hunk 4]
```diff
@@ -15,22 +15,28 @@ public TableCache(String maxwellDB) {
...
+						throw new RuntimeException("Couldn't find table " + tblName + " in database " + dbName);
```
Comment (xcvecs) [inline src/main/java/com/zendesk/maxwell/replication/TableCache.java:35]: add comment ....

**Review by xcvecs — COMMENTED** [review]
add comment

**Comment by xcvecs** [comment]
> > > Okay, when you come back, please take a look at the information. I have added an ignore_missing_schema configuration to the config.properties file：
> > > #Ignore database and table schema add not configured in filter-configuration. Default false #ignore_missing_schema = false
> > > And the following judgment will be made before throwing an exception ： if (!ignoreMissingSchema || (ignoreMissingSchema && filter.includes(dbName, tblName))) image
> > 
> > 
> > why add picture?
> 
> sorry, I don't quite understand what you mean.



> > > Okay, when you come back, please take a look at the information. I have added an ignore_missing_schema configuration to the config.properties file：
> > > #Ignore database and table schema add not configured in filter-configuration. Default false #ignore_missing_schema = false
> > > And the following judgment will be made before throwing an exception ： if (!ignoreMissingSchema || (ignoreMissingSchema && filter.includes(dbName, tblName))) image
> > 
> > 
> > why add picture?
> 
> sorry, I don't quite understand what you mean.

looks very uglylooks very uglylooks very ugly

> > > Okay, when you come back, please take a look at the information. I have added an ignore_missing_schema configuration to the config.properties file：
> > > #Ignore database and table schema add not configured in filter-configuration. Default false #ignore_missing_schema = false
> > > And the following judgment will be made before throwing an exception ： if (!ignoreMissingSchema || (ignoreMissingSchema && filter.includes(dbName, tblName))) image
> > 
> > 
> > why add picture?
> 
> sorry, I don't quite understand what you mean.

l

> > > Okay, when you come back, please take a look at the information. I have added an ignore_missing_schema configuration to the config.properties file：
> > > #Ignore database and table schema add not configured in filter-configuration. Default false #ignore_missing_schema = false
> > > And the following judgment will be made before throwing an exception ： if (!ignoreMissingSchema || (ignoreMissingSchema && filter.includes(dbName, tblName))) image
> > 
> > 
> > why add picture?
> 
> sorry, I don't quite understand what you mean.

this looks so uglylooks very ugly

> > > Okay, when you come back, please take a look at the information. I have added an ignore_missing_schema configuration to the config.properties file：
> > > #Ignore database and table schema add not configured in filter-configuration. Default false #ignore_missing_schema = false
> > > And the following judgment will be made before throwing an exception ： if (!ignoreMissingSchema || (ignoreMissingSchema && filter.includes(dbName, tblName))) image
> > 
> > 
> > why add picture?
> 
> sorry, I don't quite understand what you mean.

this looks so ugly

**Comment by xcvecs** [comment]
> > > Okay, when you come back, please take a look at the information. I have added an ignore_missing_schema configuration to the config.properties file：
> > > #Ignore database and table schema add not configured in filter-configuration. Default false #ignore_missing_schema = false
> > > And the following judgment will be made before throwing an exception ： if (!ignoreMissingSchema || (ignoreMissingSchema && filter.includes(dbName, tblName))) image
> > 
> > 
> > why add picture?
> 
> sorry, I don't quite understand what you mean.

this looks so ugly

**Comment by leehand** [comment]
> > > > Okay, when you come back, please take a look at the information. I have added an ignore_missing_schema configuration to the config.properties file：
> > > > #Ignore database and table schema add not configured in filter-configuration. Default false #ignore_missing_schema = false
> > > > And the following judgment will be made before throwing an exception ： if (!ignoreMissingSchema || (ignoreMissingSchema && filter.includes(dbName, tblName))) image
> > > 
> > > 
> > > why add picture?
> > 
> > 
> > sorry, I don't quite understand what you mean.
> 
> this looks so ugly

 Thanks for  guide, I'll change the pictures later

**Comment by osheroff** [comment]
Please add some documentation to docs/deployment.md describing the scenario you have and why you might want to use this flag -- I don't want people turning it on unless they *really* need to.

**Comment by leehand** [comment]
OK，hello!
The previous pull request was closed, so create a new pull request .
https://github.com/zendesk/maxwell/pull/1908

**Pull request closed without merging**